# mstocirc-master

## ONE. Usage description
usage: python3 analyze-tool4.py [-h] -p PEPTIDE [-s SEQUENCE] [-j JUNCTION] [-i INFO] [--output OUTPUT] [--annote ANNOTE] [--genome GENOME] [--transcript TRANSCRIPT]<br>
### For example:
There exist three models for mstocirc to analyse statistics, and below the examples are shown how to set paramaters: <br>
~$ python3 analyze-tool4.py -s seqence.fasta -p peptide.txt -j junction.fasta -o . <br>
~$ python3 analyze-tool4.py -i circRNA\_info3.txt -g GCF_000001735.4\_TAIR10.1\_genomic.fna -a Arabidopsis\_thaliana.TAIR10.47.gtf -p peptide2.txt -j junction2.fasta -o ./project <br>
~$ python3 analyze-tool4.py -i circRNA\_info3.txt -g GCF\_000001735.4\_TAIR10.1\_genomic.fna -r GCF\_000001735.4\_TAIR10.1\_transcript.fna -p peptide2.txt -j junction2.fasta

>
## TWO. Input files format description

It's necessary that the files has the same format as it can be recoginized,otherwise the program will exit. Before you run, please refer to the files in ./data and understand what formats you files need be changed into.  But don't worrry, it's not difficult to do that, and some files are listed below.<br>
 + peptide.txt is the result, generated by LC-MS/MS software analyse;<br>
 + junction.fasta is translated by 200nt spanning BSJ sequence by 6-frame;<br>
 + circRNA_info.txt is downloaded from different online databse, which demenostrates the detailed information of different circRNAs;<br>
 + sequnce.fasta, circRNA nucleotide sequnce file.<br>

see folder 'data/' for examples,
 
>
## THREE. Function presention
    
project21-10-14.1/


<table>
<tr>
<td>feature </td><td>function name </td><td> function description</td><td> folder name</td>
</tr>
<tr>
<td> m1 </td> <td> map_corf</td><td> map MS-based peptides on circRNA ORF </td><td>temp_mapp_corf </td>
</tr>
<tr> 
<td> m2 </td><td> sklearn_codinng </td><td>assess the circRNA-derived peptides </td><td>temp_skcoding</td>
</tr>
<tr>
<td> m3 </td><td> map_junct </td><td> screen MS-based peptide spanning BSJ</td><td> temp_mapp_junct</td>
</tr>
<tr>
<td> m4 </td><td> peptide_merge </td><td> merge the overlapping MS-based peptide          </td><td>temp_peptide_merge</td>
</tr>
<tr>
<td> m5 </td><td> ires_predict</td><td> predict all the possible IRES elements </td><td> temp_ires_predict</td>
</tr>
<tr>
<td> m6 </td><td> path_analysis </td><td> GO enrichment and KEGG pathways </td><td> temp_enrich</td>
</tr>
<tr>
<td> m7 </td><td> draw_circ </td><td> draw picture of the circRNAs </td><td> temp-draw_circ</td>
<tr>
<td> o1 </td><td> rem_peptide </td><td> remove MS-based repeat peptide </td><td> temp_option</td>
</tr>    
<tr>
<td> o2 </td><td> circ_classify </td><td> classify the circRNAs into six groups </td><td> temp_option</td>
</tr>
<tr>
<td>o3</td><td> circ_annote </td><td>annote the circRNA by host gene function </td><td>temp_option</td>
</tr>
<tr>
<td>o4</td><td> ms_ribo </td><td> add directive evidence from Ribo-seq analysis </td><td> temp_option </td>                           
</tr>
<tr><td> o5 </td><td> map_gene </td><td> map the MS-based peptide onto the linear transcript proteins </td><td> temp_option </td>
</tr>
</table>
Finally, generate './result/' folder.

 
>
## FOUR. Install requirement:
Mstocirc is available at https://github.com/QUMU00/mstocirc-master/, and **it can work well without being installed** after you set the parameters properly, saving the time for installing in the operating system and avoiding to be installed unsuccessfully.<br>
Mstocirc includes one main program named analze-tool4.py and several mini programs ,each of which can achieve different and important functions, and during statistics analysis, those mini programs are called to perform their specific functions by main program. Except enrichment analysis part is dependent on package 'clusterProfiler' in R environment, the bodies of other main and small are programmed by python3 language meaning that *python3 installation for your computer is necessary condition to run mstocirc*. If you want the tool to perform GO term and KEGG richment analysis, R is suggested to be installed together. But both python3 and R are the well-used programming languages for science research, therefore, it's easy to satisfy the requirements for mstocirc to run successfully.<br>
Mstocirc doesn't rely on other tool (e.g. bedtool), which is one of its advantages, but like other application developed by python3, *it needs some dependencies,such as 'matplotlib','scikit-learn' and 'numpy',etc.  modules for python3 and key package 'clusterProfiler' for R*. Users can install all the dependencies before running mstocirc or tempt to run the tool with examples input files in'/ data'. And the tool will try at first to install all of them automatically, and during this period, please keep the Internet connected.But usually, as we all know, it's better to manually install allof them by users. <br>
We introduce IRESfinder (https://github.com/xiaofengsong/IRESfinder) into mstocirc for its function IRES_predict, and here we emphasize it on purpose out of the thought that we should respect for their team copyright.<br> 
please transform your important input file format as same as example files own saved in '/data' . And it's fundamental conditions that the files you give can be recognized well by mstocirc, therefore, before running, please read the files saved in '/data' carefuuly and understand its statistics format. There are not specific requirements for general input file ï¼Œsuch as genome sequences file (.fa, .fna format) and genome annotation file (.gtf or .gff format), but we recommend to download them from NCBI website. <br>
To perform one specific function of small program, here we do not describe them about how to use in detail, and README file in their own directory are available which contain the usage information.<br>
Thank you for your downloading and running this tool.If you have any questions, plaese contact with us by email *(glli@snnu.edu.cn)*.

Edited by Zhou Cao on October 4th, 2021.<br>
Thank you....



